File: src/LanAudioRelay/Services/ReceiverSession.cs
using System.Net.Sockets;
using LanAudioRelay.Audio;
using LanAudioRelay.Core.Audio;
using LanAudioRelay.Core.Networking;
using LanAudioRelay.Core.Protocol;

namespace LanAudioRelay.Services;

public sealed class ReceiverSession : IAsyncDisposable
{
    private readonly object _gate = new();
    private readonly string _receiverId = Guid.NewGuid().ToString("N");
    private readonly AudioPlaybackSink _playback = new();
    private readonly JitterBuffer _jitterBuffer = new(targetDepth: 3, maxFrames: 128);
    private readonly short[] _silence = new short[AudioConstants.PcmSamplesPerFrame];
    private CancellationTokenSource? _cts;
    private ControlServer? _controlServer;
    private DiscoveryResponder? _discoveryResponder;
    private OpusFrameDecoder _decoder = new();
    private UdpClient? _udp;
    private Guid _activeSessionId;
    private bool _running;
    private volatile bool _hasReceivedAudio;
    private volatile bool _playoutStarted;
    private DateTimeOffset _lastDecodeErrorStatus = DateTimeOffset.MinValue;

    public event EventHandler<string>? StatusChanged;
    public event EventHandler<int>? BufferFramesChanged;
    public event EventHandler<string>? PairingCodeChanged;

    public string PairingCode { get; private set; } = "";

    public float Volume
    {
        get => _playback.Volume;
        set => _playback.Volume = value;
    }

    public Task StartAsync(CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            if (_running)
            {
                throw new InvalidOperationException("Receiver is already running.");
            }

            _running = true;
            _activeSessionId = Guid.Empty;
            _hasReceivedAudio = false;
            _playoutStarted = false;
            _decoder = new OpusFrameDecoder();
            _jitterBuffer.Reset();
            PairingCode = Pairing.GenerateCode();
            _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        }

        PairingCodeChanged?.Invoke(this, PairingCode);
        _playback.Start();

        _controlServer = new ControlServer(() => PairingCode);
        _controlServer.PairingAccepted += OnPairingAccepted;
        _controlServer.Start();

        _discoveryResponder = new DiscoveryResponder(() => new DiscoveryResponderInfo(
            _receiverId,
            Environment.MachineName,
            Environment.MachineName));
        _discoveryResponder.Start();

        _udp = new UdpClient(AudioConstants.MediaPort);
        _ = Task.Run(() => ReceiveLoopAsync(_cts!.Token), _cts.Token);
        _ = Task.Run(() => PlaybackLoopAsync(_cts!.Token), _cts.Token);

        StatusChanged?.Invoke(this, "Receiver is listening.");
        return Task.CompletedTask;
    }

    private void OnPairingAccepted(object? sender, PairingAcceptedEventArgs e)
    {
        lock (_gate)
        {
            _activeSessionId = e.SessionId;
            _hasReceivedAudio = false;
            _playoutStarted = false;
            _decoder = new OpusFrameDecoder();
            _jitterBuffer.Reset();
        }

        _playback.Clear();
        StatusChanged?.Invoke(this, $"Paired with {e.ClientName}. Waiting for audio...");
    }

    private async Task ReceiveLoopAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested && _udp is not null)
        {
            try
            {
                var result = await _udp.ReceiveAsync(cancellationToken).ConfigureAwait(false);
                if (!AudioPacket.TryParse(result.Buffer, out var packet) || packet is null)
                {
                    continue;
                }

                Guid activeSession;
                lock (_gate)
                {
                    activeSession = _activeSessionId;
                }

                if (packet.SessionId != activeSession || packet.Codec != AudioConstants.CodecOpus)
                {
                    continue;
                }

                _jitterBuffer.Push(packet);
                BufferFramesChanged?.Invoke(this, _jitterBuffer.BufferedFrameCount);

                if (!_hasReceivedAudio)
                {
                    _hasReceivedAudio = true;
                    StatusChanged?.Invoke(this, "Receiving audio packets. Playing...");
                }
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (ObjectDisposedException)
            {
                break;
            }
            catch (SocketException ex)
            {
                StatusChanged?.Invoke(this, $"UDP receive error: {ex.Message}");
            }
        }
    }

    private async Task PlaybackLoopAsync(CancellationToken cancellationToken)
    {
        using var timer = new PeriodicTimer(TimeSpan.FromMilliseconds(AudioConstants.FrameDurationMs));

        try
        {
            while (await timer.WaitForNextTickAsync(cancellationToken).ConfigureAwait(false))
            {
                if (!HasActiveSession() || !_hasReceivedAudio)
                {
                    continue;
                }

                if (!_playoutStarted && _jitterBuffer.BufferedFrameCount < 3)
                {
                    continue;
                }

                if (_jitterBuffer.TryPop(out var packet) && packet is not null)
                {
                    try
                    {
                        short[] pcm;
                        lock (_gate)
                        {
                            pcm = _decoder.Decode(packet.Payload);
                        }

                        _playback.Write(pcm);
                        _playoutStarted = true;
                    }
                    catch (Exception ex)
                    {
                        ReportDecodeError(ex);
                        _playback.Write(_silence);
                    }
                }
                else
                {
                    _playback.Write(_silence);
                }

                BufferFramesChanged?.Invoke(this, _jitterBuffer.BufferedFrameCount);
            }
        }
        catch (OperationCanceledException)
        {
        }
    }

    private bool HasActiveSession()
    {
        lock (_gate)
        {
            return _activeSessionId != Guid.Empty;
        }
    }

    private void ReportDecodeError(Exception ex)
    {
        var now = DateTimeOffset.UtcNow;
        if (now - _lastDecodeErrorStatus < TimeSpan.FromSeconds(2))
        {
            return;
        }

        _lastDecodeErrorStatus = now;
        StatusChanged?.Invoke(this, $"Audio decode error: {ex.Message}");
    }

    public async Task StopAsync()
    {
        lock (_gate)
        {
            _running = false;
            _activeSessionId = Guid.Empty;
        }

        _cts?.Cancel();
        _udp?.Dispose();
        _udp = null;

        if (_controlServer is not null)
        {
            await _controlServer.DisposeAsync().ConfigureAwait(false);
            _controlServer = null;
        }

        if (_discoveryResponder is not null)
        {
            await _discoveryResponder.DisposeAsync().ConfigureAwait(false);
            _discoveryResponder = null;
        }

        _playback.Dispose();
        _cts?.Dispose();
        _cts = null;
        PairingCode = "";
        PairingCodeChanged?.Invoke(this, PairingCode);
        BufferFramesChanged?.Invoke(this, 0);
        StatusChanged?.Invoke(this, "Receiver stopped.");
    }

    public async ValueTask DisposeAsync()
    {
        await StopAsync().ConfigureAwait(false);
    }
}

File: mac/LanAudioRelayMac/Sources/LanAudioRelayMac/ReceiverSession.swift
import Foundation
import LanAudioRelayMacCore

final class ReceiverSession {
    private let lock = NSLock()
    private let playback = MacAudioPlaybackSink()
    private let jitterBuffer = JitterBuffer(targetDepth: 3, maxFrames: 128)
    private let silence = [Int16](repeating: 0, count: AudioConstants.pcmSamplesPerFrame)
    private let receiverId = UUID().uuidString
    private var pairingCode = ""
    private var controlServer: ControlServer?
    private var discoveryResponder: DiscoveryResponder?
    private var mediaReceiver: MediaReceiver?
    private var decoder: OpusFrameDecoder?
    private var activeSessionId: UUID?
    private var playbackTask: Task<Void, Never>?
    private var running = false
    private var hasReceivedAudio = false
    private var playoutStarted = false
    private var lastDecodeErrorStatus = Date.distantPast

    var volume: Float {
        get { playback.volume }
        set { playback.volume = newValue }
    }

    var onStatus: (@Sendable (String) -> Void)?
    var onPairingCode: (@Sendable (String) -> Void)?
    var onBufferFrames: (@Sendable (Int) -> Void)?

    func start() throws {
        lock.lock()
        if running {
            lock.unlock()
            throw NSError(domain: "LanAudioRelayMac", code: 1, userInfo: [NSLocalizedDescriptionKey: "Receiver is already running."])
        }

        running = true
        activeSessionId = nil
        hasReceivedAudio = false
        playoutStarted = false
        pairingCode = Pairing.generateCode()
        decoder = try OpusFrameDecoder()
        jitterBuffer.reset()
        lock.unlock()

        try playback.start()
        onPairingCode?(pairingCode)

        let controlServer = ControlServer { [weak self] in
            self?.currentPairingCode() ?? ""
        }
        controlServer.onPairingAccepted = { [weak self] sessionId, clientName in
            self?.acceptPairing(sessionId: sessionId, clientName: clientName)
        }
        try controlServer.start()
        self.controlServer = controlServer

        let discoveryResponder = DiscoveryResponder(receiverId: receiverId) {
            Host.current().localizedName ?? "Mac"
        }
        try discoveryResponder.start()
        self.discoveryResponder = discoveryResponder

        let mediaReceiver = MediaReceiver()
        mediaReceiver.onPacket = { [weak self] packet in
            self?.receive(packet)
        }
        mediaReceiver.onError = { [weak self] message in
            self?.onStatus?("UDP receive error: \(message)")
        }
        try mediaReceiver.start()
        self.mediaReceiver = mediaReceiver

        playbackTask = Task.detached { [weak self] in
            await self?.playbackLoop()
        }

        onStatus?("Receiver is listening.")
    }

    private func currentPairingCode() -> String {
        lock.lock()
        defer { lock.unlock() }
        return pairingCode
    }

    private func acceptPairing(sessionId: UUID, clientName: String) {
        lock.lock()
        activeSessionId = sessionId
        hasReceivedAudio = false
        playoutStarted = false
        decoder = try? OpusFrameDecoder()
        jitterBuffer.reset()
        lock.unlock()

        playback.clear()
        onStatus?("Paired with \(clientName). Waiting for audio...")
    }

    private func receive(_ packet: AudioPacket) {
        lock.lock()
        let activeSessionId = self.activeSessionId
        lock.unlock()

        guard packet.sessionId == activeSessionId,
              packet.codec == AudioConstants.codecOpus
        else {
            return
        }

        jitterBuffer.push(packet)
        onBufferFrames?(jitterBuffer.bufferedFrameCount)

        lock.lock()
        let firstAudio = !hasReceivedAudio
        hasReceivedAudio = true
        lock.unlock()

        if firstAudio {
            onStatus?("Receiving audio packets. Playing...")
        }
    }

    private func playbackLoop() async {
        while !Task.isCancelled {
            try? await Task.sleep(nanoseconds: UInt64(AudioConstants.frameDurationMs) * 1_000_000)

            lock.lock()
            let hasSession = activeSessionId != nil
            let shouldRead = hasSession && hasReceivedAudio
            let shouldPrime = !playoutStarted && jitterBuffer.bufferedFrameCount < 3
            lock.unlock()

            guard shouldRead, !shouldPrime else {
                continue
            }

            if let packet = jitterBuffer.pop() {
                do {
                    lock.lock()
                    let decoder = self.decoder
                    lock.unlock()

                    guard let decoder else {
                        continue
                    }

                    let pcm = try decoder.decode(packet.payload)
                    playback.write(pcm)

                    lock.lock()
                    playoutStarted = true
                    lock.unlock()
                } catch {
                    reportDecodeError(error)
                    playback.write(silence)
                }
            } else {
                playback.write(silence)
            }

            onBufferFrames?(jitterBuffer.bufferedFrameCount)
        }
    }

    private func reportDecodeError(_ error: Error) {
        let now = Date()
        guard now.timeIntervalSince(lastDecodeErrorStatus) > 2 else {
            return
        }
        lastDecodeErrorStatus = now
        onStatus?("Audio decode error: \(error.localizedDescription)")
    }

    func stop() {
        lock.lock()
        running = false
        activeSessionId = nil
        pairingCode = ""
        hasReceivedAudio = false
        playoutStarted = false
        lock.unlock()

        playbackTask?.cancel()
        playbackTask = nil
        mediaReceiver?.stop()
        mediaReceiver = nil
        discoveryResponder?.stop()
        discoveryResponder = nil
        controlServer?.stop()
        controlServer = nil
        playback.stop()
        jitterBuffer.reset()

        onPairingCode?("")
        onBufferFrames?(0)
        onStatus?("Receiver stopped.")
    }
}
